File: M80AttributedLabel/OCDemo/Classes/DemoTableViewController.m
//
//  DemoTableViewController.m
//  M80AttributedLabel
//
//  Created by amao on 4/21/14.
//  Copyright (c) 2014 www.xiangwangfeng.com. All rights reserved.
//

#import "DemoTableViewController.h"

@interface TableItem : NSObject
@property (nonatomic,strong)    NSString    *title;
@property (nonatomic,strong)    NSString    *subTitle;
@property (nonatomic,strong)    NSString    *vcName;
@end

@implementation TableItem
+ (TableItem *)itemWithTitle:(NSString *)title
                    subTitle:(NSString *)subTitle
                      vcName:(NSString *)vcName
{
    TableItem *instance = [[TableItem alloc]init];
    instance.title      = title;
    instance.subTitle   = subTitle;
    instance.vcName= vcName;
    return instance;
}
@end

@interface DemoTableViewController ()
@property (nonatomic,strong)    NSArray *items;
@end

@implementation DemoTableViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.title = @"M80AttributedLabel";
    self.navigationController.view.backgroundColor = [UIColor whiteColor];
    
    _items = @[[TableItem itemWithTitle:@"Basic"
                               subTitle:@"How to create a simple M80AttributedLabel"
                                 vcName:@"BasicAttributedLabelViewController"],
               
               
               [TableItem itemWithTitle:@"Customizing Text"
                               subTitle:@"How to use NSAttributedString"
                                 vcName:@"CustomTextViewController"],
               
               [TableItem itemWithTitle:@"Links"
                               subTitle:@"Automatic and explicit Links"
                                 vcName:@"LinksViewController"],
               
               [TableItem  itemWithTitle:@"Images"
                                subTitle:@"How to display images using M80AttributedLabel"
                                  vcName:@"ImagesViewController"],
               
               [TableItem itemWithTitle:@"LongText"
                               subTitle:@"Performance for displaying long text"
                                 vcName:@"LongTextViewController"],
               
               [TableItem itemWithTitle:@"Attachment"
                               subTitle:@"Add UIView as attachment instead of UIImage"
                                 vcName:@"AttachmentViewController"],
               
               [TableItem itemWithTitle:@"LineBreakMode"
                               subTitle:@"Different line break modes"
                                 vcName:@"LineBreakModeViewController"],
               
               [TableItem itemWithTitle:@"TextAlignment"
                               subTitle:@"Different test alignments"
                                 vcName:@"TextAlignmentViewController"],
               
               [TableItem itemWithTitle:@"TableViewCell"
                               subTitle:@"AttributedLabel in tableviewcell"
                                 vcName:@"TextTableViewController"],
               
               ];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
}

#pragma mark - Table view data source
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return [_items count];
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"test_demo_cell"];
    if (cell == nil)
    {
        cell = [[UITableViewCell alloc]initWithStyle:UITableViewCellStyleSubtitle
                                     reuseIdentifier:@"test_demo_cell"];
    }
    TableItem *item = [_items objectAtIndex:[indexPath row]];
    cell.textLabel.text = item.title;
    cell.detailTextLabel.text = item.subTitle;
    return cell;
}



#pragma mark - Table view delegate
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{

    TableItem *item = [_items objectAtIndex:[indexPath row]];
    NSString *vcName= [item vcName];
    UIViewController *controller = [[NSClassFromString(vcName) alloc] init];
    controller.title = [item title];
    [self.navigationController pushViewController:controller animated:YES];
}


@end

File: SwiftDemo/Classes/DemoTableViewController.swift
//
//  DemoTableViewController.swift
//  SwiftDemo
//
//  Created by amao on 2016/10/12.
//  Copyright © 2016年 amao. All rights reserved.
//

import UIKit

struct Item {
    let title : String
    let subTitle : String
    let vcName : String
    
    init(_ title : String, _ subTitle : String, _ vcName : String) {
        self.title = title
        self.subTitle = subTitle
        self.vcName = vcName
    }
}




class DemoTableViewController: UITableViewController {
    
    let items : [Item] = [Item("Basic","How to create a simple M80AttributedLabel","BasicAttributedLabelViewController"),
                          Item("Customizing Text","How to use NSAttributedString","CustomTextViewController"),
                          Item("Links","Automatic and explicit Links","LinksViewController"),
                          Item("Images","How to display images using M80AttributedLabel","ImagesViewController"),
                          Item("LongText","Performance for displaying long text","LongTextViewController"),
                          Item("Attachment","Add UIView as attachment instead of UIImag","AttachmentViewController"),
                          Item("LineBreakMode","Different line break modes","LineBreakModeViewController"),
                          Item("TextAlignment","Different test alignments","TextAlignmentViewController")]

    override func viewDidLoad() {
        super.viewDidLoad()
        self.title = "Demo"
        self.navigationController?.view.backgroundColor = UIColor.white
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }

    // MARK: - Table view data source
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.items.count
    }

    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        let identifier = "demo_identifier"
        let item = self.items[indexPath.row]
        var cell = tableView.dequeueReusableCell(withIdentifier:identifier)
        if cell == nil {
            cell = UITableViewCell(style: .subtitle, reuseIdentifier: identifier)
        }
        cell?.textLabel?.text = item.title
        cell?.detailTextLabel?.text = item.subTitle

        return cell!
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let item = self.items[indexPath.row]
        let name = item.vcName
        if let vcClass = ClassByName(name: name) as? UIViewController.Type {
            let vc = vcClass.init()
            self.navigationController?.pushViewController(vc, animated: true)
        }
        
    }

}
